File: leetcode/algorithms/203_remove_linked_list_elements/main.dart
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class RemovrlinkedListElements {
  /**
   * Dummy Node
   *
   * Complexities:
   *   N - Number of nodes in `head`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  ListNode? removeElements(ListNode? head, int val) {
    ListNode? result = ListNode();
    result.next = head;
    ListNode? current = result;

    while (current?.next != null) {
      if (current?.next?.val == val) {
        current?.next = current.next?.next;
      } else {
        current = current?.next;
      }
    }

    return result.next;
  }


  // Solution
  /**
   * Dummy Node
   *
   * Complexities:
   *   N - Number of nodes in `head`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  ListNode? solution(ListNode? head, int val) {
    ListNode dummy = ListNode(0, head);
    ListNode current = dummy;

    while (current.next != null) {
      ListNode nextNode = current.next!;

      if (nextNode.val == val) {
        current.next = nextNode.next;
      } else {
        current = nextNode;
      }
    }

    return dummy.next;
  }
}

File: leetcode/algorithms/203_remove_linked_list_elements/main.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * Dummy Node
 *
 * Complexities:
 *   N - Number of nodes in `head`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function (head, val) {
  if (!head) {
    return null;
  }

  let currentHead = head;
  let nextHead = head.next;

  while (nextHead) {
    if (nextHead.val === val) {
      nextHead = nextHead.next;
      currentHead.next = nextHead;
    } else {
      currentHead = nextHead;
      nextHead = nextHead.next;
    }
  }

  if (head.val === val) {
    return head.next;
  }

  return head;
};


// Solution
/**
 * Dummy Node
 *
 * Complexities:
 *   N - Number of nodes in `head`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var solution = function (head, val) {
  const dummy = new ListNode(0, head);

  let current = dummy;

  while (current.next !== null) {
    if (current.next.val === val) {
      current.next = current.next.next;
    } else {
      current = current.next;
    }
  }

  return dummy.next;
};
